src/main/java/com/icthh/xm/tmf/ms/communication/service/firebase/FirebaseService.java

```diff
@@ -146,30 +130,38 @@ private FirebaseMessaging getFirebaseMessaging(CommunicationMessage message) {
             .orElseThrow(() -> new BusinessException("error.fcm.sender.id.invalid", "Sender id is not valid"));
     }
 
-    private CommunicationMessage buildCommunicationResponse(BatchResponse batchResponse, CommunicationMessage message) {
+    private CommunicationMessage buildCommunicationResponse(BatchResponse batchResponse, CommunicationMessage msg,
+                                                            List<Receiver> receivers) {
+        ExtendedCommunicationMessage message = ExtendedCommunicationMessageFactory.newMessage(msg);
 
+        List<Detail> details = new ArrayList<>();
         List<SendResponse> responses = batchResponse.getResponses();
-        List<CommunicationRequestCharacteristic> characteristics = new ArrayList<>(responses.size());
-        characteristics.add(new CommunicationRequestCharacteristic().name("successCount").value(String.valueOf(batchResponse.getSuccessCount())));
-        characteristics.add(new CommunicationRequestCharacteristic().name("failureCount").value(String.valueOf(batchResponse.getFailureCount())));
+        for (int i = 0; i < responses.size(); i++) {
+            SendResponse response = responses.get(i);
+
+            Detail detail = new Detail()
+                .receiver(receivers.get(i));
 
-        for (SendResponse response : responses) {
-            String rName, rValue;
             if (response.isSuccessful()) {
-                rName = "messageId";
-                rValue = response.getMessageId();
+                detail.status(Detail.Status.SUCCESS)
+                    .messageId(response.getMessageId());
             } else {
-                rName = "error";
-                rValue = response.getException().getMessage();
+                FirebaseMessagingException exception = response.getException();
+
+                detail.status(Detail.Status.ERROR)
+                    .error(new ErrorDetail()
+                        .code(String.valueOf(exception.getMessagingErrorCode()))
+                        .description(exception.getMessage()));
             }
-            characteristics.add(new CommunicationRequestCharacteristic().name(rName).value(rValue));
+            details.add(detail);
```

Comment:
Do we really need unconditional details adding to batch response? I mean in case all messages were success we will get useless but big output. 
Propose considering applying different strategies to response preparation:
1. DETAIL_FULL - responds with full details does not matter of resp code (as it is now)
2. DETAIL_SUMMARY - responds only with counts (success vs failures) - propose to make it default
3. DETAIL_ERROR - returns detail only for non success records.

maybe there are also other options.

Reply:
Good point, thanks.
Do we want to pass the expected detail strategy with a request parameter or have an application-wide configuration?

Reply:
@amedvedchuk I've added the strategies that are passed as a request parameter.